wolfcrypt/src/wc_pkcs11.c

```diff
@@ -148,25 +150,36 @@ void wc_Pkcs11_Finalize(Pkcs11Dev* dev)
 int wc_Pkcs11Token_Init(Pkcs11Token* token, Pkcs11Dev* dev, int slotId,
     const char* tokenName, const unsigned char* userPin, int userPinSz)
 {
-    int        ret = 0;
-    CK_RV      rv;
-    CK_SLOT_ID slot;
-    CK_ULONG   slotCnt = 1;
+    int         ret = 0;
+    CK_RV       rv;
+    CK_SLOT_ID* slot = NULL;
+    CK_ULONG    slotCnt = 0;
 
     if (token == NULL || dev == NULL || tokenName == NULL)
         ret = BAD_FUNC_ARG;
 
     if (ret == 0) {
         if (slotId < 0) {
             /* Use first available slot with a token. */
-            rv = dev->func->C_GetSlotList(CK_TRUE, &slot, &slotCnt);
+            rv = dev->func->C_GetSlotList(CK_TRUE, NULL, &slotCnt);
             if (rv != CKR_OK)
                 ret = WC_HW_E;
+            if (ret == 0) {
+                slot = (CK_SLOT_ID*)XMALLOC(slotCnt * sizeof(*slot), NULL,
```

Comment:
Let's add a dev->heap member that gets set during the device init() for all XMALLOCs and FREEs.   Much easier to do now and maintain then to try and add later with the interfaces already out there.